### ISSUE TITLE
Function returning a switch expression returns an Object class instead of a class type

### ISSUE THREAD

**Issue by justWalsdi** [dart-lang/sdk#53244]
When you create a function that returns a `switch` expression that returns an instance of a class that inherits from some `abstract class BaseClass<Object>` AND specifies the type after the inherited class(`class F extends BaseClass<Object>`), then the return type of the function turns into an `Object`. The `runtimeType` of these objects is still `BaseClass`. It can be fixed by turning switch expression into a switch statement.

It can be reproduced with the following code:
```dart
abstract class BaseClass<Object> { BaseClass(); }

class F extends BaseClass<Object> { F(); }

class S extends BaseClass { S(); }

enum V { f, s }

// Works perfectly fine
BaseClass switchStatement(V v) {
  switch (v) {
    case V.f: return F();
    case V.s: return S();
  }
}

// Causes an error:
// Error: A value of type 'Object' can't be returned
// from a function with return type 'BaseClass<Object>'.
BaseClass switchExpression(V v) => switch (v) {
      V.f => F(),
      V.s => S(),
    };

void main() {}
```

The workaround is using `dynamic` as a return type or add `as BaseClass` after the class constructor call. The first solution communicates a wrong idea. The second solution confuses Dart Analysis server because it is an unnecessary cast. 

What am I missing?

This tracker is for issues related to:

* dev_compiler
* Dart SDK version: `3.0.6 (stable)`

**Comment by lrhn** [comment]
That's just how the you bound of `BaseClass<Object> ` and `BaseClass<dynamic>` is computed.

Your two branches have different types.
To figure out which type to give to the switch, which can evaluate to either branch, the compiler tries to find a type that both branches satisfy, an upper bound on the types of all the branches.
The easy that works it's that it looks at the interfaces that both types have, but since they implement different versions of `BaseClass`, that's not seen as a solution, so the closest shared interface is `Object`.

It's a known issue that guessing at the upper bound is not particularly smart.
There are hard cases where a smart algorithm can get lost, but the current algorithm may be erring on the side of dumbness a little too much.

**Comment by justWalsdi** [comment]
Hi, thank you for your reply. When I encountered the problem, `BaseClass` was defined as `class BaseClass<T extends Object>`. I tried to make the simplest example with which this problem can be reproduced. If I change the code to the following:

```dart
abstract class BaseClass<T extends Object> { BaseClass(); }

class F extends BaseClass<Object> { F(); }

class S extends BaseClass { S(); }
```

This works, but only if the `F` class type parameter is `Object`. Anything extending `Object` in a type parameter will cause the switch expression to guess the return type as `Object` once again.

Either way, I now understand why it is happening and will close this issue. Thanks for the clarification.
